examples/jvm-interop/README.md

```diff
@@ -0,0 +1,210 @@
+# JVM interop
+This is a demo for calling Roc code from Java, and some other JVM languages.
+
+
+## Prerequisites
+
+The following was tested on NixOS, with `openjdk 17.0.5` and `clang 13.0.1` but should work with most recent versions of those (jdk>=10) on most modern Linux and MacOS.\
+You're welcome to test on your machine and tell me (via [Zulip](https://roc.zulipchat.com/#narrow/pm-with/583319-dank)) if you ran into any issues or limitations.
+
+## Goal
+We'll make a few examples, showing basic data type convertions and function calls between Roc and Java (and later some other JVM languages):
+- A string formatter.
+- A Function that multiples an array by a scalar.
+- A factorial function that, for the sake of demonstration, throws a RuntimeException for negative integers.
+
+This will be done with the help of [Java Native Interface](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/).
+We will be using C to bridge between Java and Roc.
+
+## Structure
+As the time of writing this post, the following is the current bare bones tree of a jvm-interop:
+
+``` console
+.
+├── impl.roc
+├── platform.roc
+├── bridge.c
+└── javaSource
+    └── Demo.java
+```
+
+impl.roc is the application where we actually implement our native Roc functions.\
+platform.roc as the name suggests contains platform logic, (but doesn't really have much here, mostly just) exposes functions to the host - bridge.c\
+bridge.c is the JNI bridge, it's the host that implements the Roc functions (e.g roc_alloc) and the JNI functions that act like the bridge between Roc and Java (bridge as in, doing type conversions between the languages, needed jvm boilerplate, etc). 
+
+For each of our native Roc functions, in the application (impl.roc), we have a corresponding `Java_javaSource_Demo_FUNC` C function that handles the "behind the scenes", this includes type conversions between the languages, transforming roc panics into java exceptions and basically all the glue code necessary.
+
+
+Just so you know what to expect, our Roc functions look like this;
+``` coffee
+interpolateString : Str -> Str
+interpolateString = \name ->
+    "Hello from Roc \(name)!!!🤘🤘🤘"
+
+
+mulArrByScalar : List I32, I32 -> List I32
+mulArrByScalar  = \arr, scalar ->
+    List.map arr \x -> x * scalar
+
+
+factorial : I64 -> I64
+factorial = \n ->
+    if n < 0 then
+        # while we get the chance,  examplify a roc panic in an interop
+        crash "No negatives here!!!"
+    else if n == 0 then
+        1
+    else
+        n * (factorial (n - 1))
+```
+
+Nothing too crazy. Again, do note how we crash if n < 0, see how this would play out from the Java side. 
+
+Now let's take a quick look on the Java side of things;
+
+``` java
```

Comment:
I'm not sure if it is worth putting code directly in the readme. Given this is an example, it is not that likely to get stale, but as things change, it would be confusing if this differs. Instead, maybe it is better to just add a link to the code.

Reply:
i mean you're right but i wanted to explain the decls step by step so figured it's nicer if there's visuals right in front of you

Reply:
so i guess if it does change (which indeed is unlikely), ill update/ keep including only the relevant parts for the explanations, worth the possible hassle in my opinion